syntax = "proto3";

option java_multiple_files = true;

message FileMessage {
    bytes file = 1;
    repeated string options = 2;
}

message ResourceIdentifier {
    string resource_id = 1;
}

message RuntimeMessage {
    string resource_id = 1;
    repeated string property = 2;
    bytes file = 3;
}

message StringResponse {
    string response = 1;
}

message StartStopMessage{
    string container_id = 1;
}

message Empty {

}

message Auth {
        string auth_url = 1;
        string username = 2;
        string password = 3;
        string project = 4;
}

message ResourceGroupProto {

    message PoP {
        string name = 1;
        string interfaceEndpoint = 2;
    }

    message MetadataEntry {
        string key = 1;
        string value = 2;
    }

    message Network {
        string name = 1;
        string poPName = 2;
        string cidr = 3;
        string networkId = 4;
    }

    message VDU {
        string name = 1;
        string imageName = 2;
        string netName = 3;
        string poPName = 4;
        string computeId = 5;
        string ip = 6;
        repeated MetadataEntry metadata = 7;
    }

    string name = 1;
    repeated PoP pops = 2;
    repeated Network networks = 3;
    repeated VDU vdus = 4;
}

service OperationHandler {

    rpc Create(FileMessage) returns (ResourceGroupProto) {}

    rpc Remove(ResourceIdentifier) returns (Empty) {}

    rpc StopContainer(ResourceIdentifier) returns (Empty) {}

    rpc CheckIfContainerExists(ResourceIdentifier) returns (StringResponse) {}

    rpc CheckIfContainerRunning(ResourceIdentifier) returns (StringResponse) {}

    rpc StartContainer(ResourceIdentifier) returns (Empty) {}

    rpc ExecuteCommand(RuntimeMessage) returns (StringResponse) {}

    rpc DownloadFile(RuntimeMessage) returns (FileMessage) {}

    rpc UploadFile(RuntimeMessage) returns (Empty) {}
}
